Stats Service

aletya opened the pull request:
NOTE:
Added 
attendanceCount: {
        type: Number,
        default: 0,
},
field to EventSchema + EventValidator

Stats Service:

GET /stats/check-in
Get the number of people checked in (staff only)
-works by getting number of attendees with 1+ events
example response:
{count: 10}

GET /stats/merch-item/:PRICE
Get the number of people eligible for merch item (staff only)
-works by getting number of people with points >= PRICE
example response:
{count: 10}

GET /stats/priority-attendee
Get the number of priority attendees (staff only)
-works by getting all attendees with priority-expiration field in the future
example response:
{count: 10}

GET /stats/attendance/:N
Get the attendance of the past n events (staff only)
-sorted in most recent first
example response:
{attendanceCounts: [2, 4, 2, 1, 10]}

GET /stats/dietary-restrictions
-gets a breakdown of all dietary-restrictions
example response: 
{dietaryRestrictions: {"None": 230, "Milk" : 10}}

divyack2 commented on src/services/stats/stats-router.ts:
```diff
@@ -0,0 +1,127 @@
+            const dietaryRestrictionsMap: { [key: string]: number } = {};
+
+            attendees.forEach((attendee) => {
+                const dietary_restriction =
+                    attendee.dietary_restrictions as string;
```
@AydanPirani question: why did we make dietary restrictions a string and not a string array?

aletya replied:
Todo: change dietary restrictions to a string array w/ enums
separate dietary restrictions and allergies
for 8 most common allergies, do boolean values (check mark)
add attendee checkin for each day (subobject, arr)

aletya replied:
and store all items in dietary restrictions / allergies array as all upper case

aletya replied:
also change endpoints in registration to match it

aletya replied:
no restrictions, has  ONLY dietary restriction, has ONLY allergies, has BOTH.
^all exclusive

divyack2 reviewed (approved):
looks good! hold off on merging though, I just wanna double-check one field :)

aletya commented:
{"none":1,"dietary_restrictions":0,"allergies":0,"both":2}

aletya commented:
Changes:
changed dietary restrictions to string[] in BOTH registration and attendees
added allergies : string[] to BOTH registration and attendees
added checkin : boolean[5] to attendees

changed GET /stats/dietary-restrictions to return data in the format:
{
  "none": 140,
  "dietary_restrictions": 5,
  "allergies": 10,
  "both": 23
}

aletya commented:
{
  "none": 1,
  "dietary_restrictions": 0,
  "allergies": 0,
  "both": 2,
  "allergy_counts": {
    "EGGS": 2,
    "MILK": 1
  },
  "dietary_restriction_counts": {
    "HALAL": 2,
    "VEGETARIAN": 2
  }
}

AydanPirani commented on src/services/attendees/attendee-schema.ts:
```diff
@@ -7,9 +7,11 @@ const AttendeeValidator = z.object({
     email: z.string().email(),
     events: z.array(z.string()),
-    dietary_restrictions: z.string(),
+    dietary_restrictions: z.string().array(),
```
follow js naming conventions 

```suggestion
    dietaryRestrictions: z.string().array(),
```

AydanPirani commented on src/services/attendees/attendee-schema.ts:
```diff
@@ -7,9 +7,11 @@ const AttendeeValidator = z.object({
+    allergies: z.string().array(),
     priority_expiry: z.date().nullable().optional(),
     points: z.number().min(0).default(0),
+    checkin: z.boolean().array().default([false, false, false, false, false]),
```
change this to hasPriority, should be a dictionary of days (or at least extend some interface for us to check if priority is given on a certain day)

AydanPirani commented on src/services/attendees/attendee-schema.ts:
```diff
@@ -18,9 +20,11 @@ const AttendeeSchema = new mongoose.Schema({
     email: { type: String, required: true, unique: true },
     events: [{ type: mongoose.Schema.Types.ObjectId, ref: "Event" }],
-    dietary_restrictions: { type: String, required: true },
+    dietary_restrictions: { type: [String], required: true },
```
js convention

AydanPirani commented on src/services/attendees/attendee-schema.ts:
```diff
@@ -18,9 +20,11 @@ const AttendeeSchema = new mongoose.Schema({
+    allergies: { type: [String], required: true },
     priority_expiry: { type: Date, default: null },
     points: { type: Number, default: 0 },
+    checkin: { type: [Boolean], deafult: [false, false, false, false, false] },
```
match above changes, default is misspelled

AydanPirani commented on src/services/events/events-schema.ts:
```diff
@@ -12,6 +12,7 @@ export const EventValidator = z.object({
     isVirtual: z.boolean(),
     imageUrl: z.string().nullable().optional(),
     isVisible: z.boolean().default(false),
+    attendanceCount: z.number().optional(),
```
since we've added metadata to events, these shoulnd't be visible in GET /event anymore. Can you filter these out from GET /event?

How i'd do it is:

- create a zod validator for publicEventData
- create a zod validator for internalEventData that extends public event Data
- in get /event, pass it through public event data to filter out the extra fields

AydanPirani commented on src/services/stats/stats-router.ts:
```diff
@@ -0,0 +1,161 @@
+        try {
+            const currentTime = new Date();
+            const attendees = await Database.ATTENDEES.find({
+                priority_expiry: { $gt: currentTime },
```
this field won't exist, u need to use the `hasCheckedIn` field

AydanPirani commented on src/services/stats/stats-router.ts:
```diff
@@ -0,0 +1,161 @@
+    RoleChecker([Role.enum.STAFF], false),
+    async (req, res, next) => {
+        try {
+            const n = req.params.N;
```
```suggestion
            const numEvents = req.params.N;
```

AydanPirani commented on src/services/stats/stats-router.ts:
```diff
@@ -0,0 +1,161 @@
+    RoleChecker([Role.enum.STAFF], false),
+    async (req, res, next) => {
+        try {
+            const attendees = await Database.ATTENDEES.find({});
```
Don't load all the attendees in ram, do it via mongo endpoints. You should be able to flatten/unwrap an array, then get counts through that.

aletya replied:
do in parallel with promise.all

aletya replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled
@alex
^^^
use allSettled instead of any/all

AydanPirani commented on src/services/stats/stats-router.ts:
```diff
@@ -0,0 +1,161 @@
+            const dietary_restriction_counts: { [key: string]: number } = {};
+            const allergy_counts: { [key: string]: number } = {};
+
+            attendees.forEach((attendee) => {
+                if (
+                    attendee.dietary_restrictions.length > 0 &&
+                    attendee.allergies.length > 0
+                ) {
+                    both++;
+                } else if (attendee.dietary_restrictions.length > 0) {
+                    dietary_restrictions++;
+                } else if (attendee.allergies.length > 0) {
+                    allergies++;
+                } else {
+                    none++;
+                }
+
+                const attendee_dietary_restrictions: string[] =
+                    attendee.dietary_restrictions;
```
no need to worry about both/none imo, just counts of most popular ones

AydanPirani replied:
and then the counts of extra ones

aletya replied:
Im j gonna leave those there cuz it might be nice to have for the admin dashboard

aletya commented:
Other / notes for alex : set up infra for testing (jest)
Also: default.env files

AydanPirani commented on src/services/events/events-schema.ts:
```diff
@@ -12,7 +12,19 @@ export const EventValidator = z.object({
-    attendanceCount: z.number().optional(),
+});
+
+export const privateEventValidator = z.object({
```
Extend public event validator, don’t create a whole new one

AydanPirani reviewed (changes requested):
Suggested changes, fix linter too pls :)

AydanPirani commented on src/services/attendees/attendee-schema.ts:
```diff
@@ -6,17 +6,43 @@ const AttendeeValidator = z.object({
+    allergies: z.string().array(),
+    hasCheckedIn: z.boolean().default(false),
     points: z.number().min(0).default(0),
+    hasPriority: z.record(z.boolean()).default({
```
use z.object instead of z.record, and pass in the keys

AydanPirani commented on src/services/events/events-router.ts:
```diff
@@ -40,15 +40,21 @@ eventsRouter.post("/", async (req, res, next) => {
 
-        return res.status(StatusCodes.OK).json(event.toObject());
+        const filtered_event = publicEventValidator.parse(
+            unfiltered_event.toJSON()
```
why ToJSON?

aletya replied:
thought i needed it but apparently not 👍

AydanPirani commented on src/services/events/events-schema.ts:
```diff
@@ -4,7 +4,7 @@ import { v4 as uuidv4 } from "uuid";
 
-export const EventValidator = z.object({
+export const publicEventValidator = z.object({
     eventId: z.coerce.string().optional(),
```
optional?

AydanPirani reviewed (approved):
lgtm, small changes but otw we good
